### PR TITLE
clone3 error check based on clone_args version

### PR DESCRIPTION
**Pull request by sv641**

### CODE REVIEW

File: km/km_init_guest.c [hunk 1]
```diff
@@ -329,14 +329,23 @@ int km_clone(km_vcpu_t* vcpu,
    return km_vcpu_get_tid(new_vcpu);
 }
 
-long km_clone3(km_vcpu_t* vcpu, struct clone_args* cl_args)
+long km_clone3(km_vcpu_t* vcpu, struct clone_args* cl_args, size_t cl_args_len)
 {
-   if (cl_args->set_tid_size != 0 || cl_args->cgroup != 0) {
+#ifdef CLONE_ARGS_SIZE_VER2
```
Comment (johnmuth81) [inline km/km_init_guest.c:334]: What about CLONE_ARGS_SIZE_VER0? That exists.

Comment (sv641) [reply km/km_init_guest.c:334]: Let me rephrase all V0 members are used.

File: km/km_init_guest.c [hunk 2]
```diff
@@ -329,14 +329,23 @@ int km_clone(km_vcpu_t* vcpu,
...
+   if ((cl_args_len >= CLONE_ARGS_SIZE_VER1 && cl_args->set_tid_size != 0) ||
+       (cl_args_len >= CLONE_ARGS_SIZE_VER2 && cl_args->cgroup != 0)) {
       km_warnx("Not supported clone3 args: tid_size, cgroup %llu, %llu",
                cl_args->set_tid_size,
                cl_args->cgroup);
       return -ENOSYS;
    }
+#elif CLONE_ARGS_SIZE_VER1
+   if (cl_args_len >= CLONE_ARGS_SIZE_VER1 && cl_args->set_tid_size != 0)
```
Comment (gnode1) [inline km/km_init_guest.c:343]: forgotten `{` and `}`

**Review by gnode1 — APPROVED** [review]

File: km/km_init_guest.c [hunk 3]
```diff
@@ -329,14 +329,23 @@ int km_clone(km_vcpu_t* vcpu,
...
-   if (cl_args->set_tid_size != 0 || cl_args->cgroup != 0) {
+#ifdef CLONE_ARGS_SIZE_VER2
+   if ((cl_args_len >= CLONE_ARGS_SIZE_VER1 && cl_args->set_tid_size != 0) ||
+       (cl_args_len >= CLONE_ARGS_SIZE_VER2 && cl_args->cgroup != 0)) {
       km_warnx("Not supported clone3 args: tid_size, cgroup %llu, %llu",
                cl_args->set_tid_size,
                cl_args->cgroup);
       return -ENOSYS;
    }
+#elif CLONE_ARGS_SIZE_VER1
```
Comment (paulpopelka) [inline km/km_init_guest.c:342]: Should this be #elif defined(CLONE_ARGS_SIZE_VER1)  ??

The check for CLONE_ARGS_SIZE_VER2 is with #ifdef

File: km/km_init_guest.c [hunk 4]
```diff
@@ -329,14 +329,23 @@ int km_clone(km_vcpu_t* vcpu,
    return km_vcpu_get_tid(new_vcpu);
 }
 
-long km_clone3(km_vcpu_t* vcpu, struct clone_args* cl_args)
+long km_clone3(km_vcpu_t* vcpu, struct clone_args* cl_args, size_t cl_args_len)
 {
-   if (cl_args->set_tid_size != 0 || cl_args->cgroup != 0) {
+#ifdef CLONE_ARGS_SIZE_VER2
+   if ((cl_args_len >= CLONE_ARGS_SIZE_VER1 && cl_args->set_tid_size != 0) ||
+       (cl_args_len >= CLONE_ARGS_SIZE_VER2 && cl_args->cgroup != 0)) {
```
Comment (paulpopelka) [inline km/km_init_guest.c:336]: cgroup is an fd, fd 0 is valid.
Seems like you should check the CLONE_INTO_CGROUP bit in the flags field to decide if they are using cgroup.

File: km/km_init_guest.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulpopelka) [reply km/km_init_guest.c:334]: I agree.

**Comment by sv641** [comment]
Hopefully I got it right this time.

File: km/km_init_guest.c (hunk 3 above)
Comment (sv641) [reply km/km_init_guest.c]: changed to using defined in both places

File: km/km_init_guest.c (hunk 4 above)
Comment (sv641) [reply km/km_init_guest.c]: changed

File: km/km_init_guest.c (hunk 2 above)
Comment (sv641) [reply km/km_init_guest.c]: fixed
